drivers/phy/cadence/phy-cadence-torrent.c

```diff
@@ -30,136 +31,163 @@
 #define POLL_TIMEOUT_US		5000
 #define LANE_MASK		0x7
 
+#define TORRENT_COMMON_CDB_OFFSET	0x0
+
+#define TORRENT_TX_LANE_CDB_OFFSET(ln, offset)	\
+				(0x4000 + ((ln) * (0x200 << (2 - (offset)))))
```

Comment:
I think this macro should reflect what is given in the user guide
Address bits 15:9 => 01_m_nnnn
Addressed Region => General lane addressing : m = multi-write, nnnn = lane number.
                                    Transceiver And Transmitter Components Address Space on page 60
We can ignore multi-write but lane_number << 9 can be done.
Do you see any issues with that being followed?

Reply:
The macro here is written considering Cadence and TI environments. For (2 - offset) part, offset will be 2 for cadence and 1 for TI. This will calculate register offsets in both environments.
For Cadence environment, as (ln * 0x200) is equivalent to (ln << 9), offsets will be calculated correctly. For TI case, we can't use (ln << 9) directly, as per my understanding. 
Do you want (0x200) should be replaced by (1 << 9) here? Please let me know if I understood you correctly and also please suggest any better alternative if you have.

Reply:
From your comment, it looks like Torrent in Cadence environment defers from Sierra in Cadence environment. In sierra Cadence environment, the entire offset should be left shifted by '2'. 
For Sierra, I'm changing the lane offset calculation to something like below
#define SIERRA_LANE_CDB_OFFSET(ln, block_offset, reg_offset)   \
+                               ((0x4000 << (block_offset)) + \
+                                (((ln) << 9) << (reg_offset)))

The block offset and reg offset above are variables and it can change based on platforms. The 0x4000 and << 9 are according to the user guide, so that shouldn't be disturbed.

The block offset above was added specifically for Cadence environment since the entire block has to be shifted by '2'.

If there's no such requirement for Torrent and the only variance is for TI platform (from the userguide), I suggest to add something like

#define  TORRENT_TX_LANE_CDB_OFFSET(ln, reg_offset)   \
+                               (0x4000 + \
+                                (((ln) << 9) << (reg_offset)))

I assume reg_offset should be '0' for Torrent in Cadence environment.

Reply:
Just to correct my comment above and clarify the things, macro TORRENT_TX_LANE_CDB_OFFSET is the block offset and that is shifted by block_offset_shift (2 for Cadence, 0 for TI) in function cdns_regmap_init(), just like you have implemented in Sierra driver. I have referred your code from 
https://lore.kernel.org/lkml/20191016113117.12370-4-kishon@ti.com/
and same is implemented in Torrent driver.
So, I think Torrent and Sierra offsets are similar in Cadence environment.

If we use macro as below just as you are saying to use in Sierra driver:
#define TORRENT_LANE_CDB_OFFSET(ln, block_offset_shift, reg_offset_shift)
((0x4000 << (block_offset_shift)) +
(((ln) << 9) << (reg_offset_shift)))

We may need to change line in function cdns_regmap_init() as:
-  ctx->base = base + (block_offset << block_offset_shift);
+ ctx->base = base;

as shifting is already done in macro TORRENT_LANE_CDB_OFFSET.
With block_offset_shift = 2 and reg_offset_shift = 2 (For Cadence) 
and block_offset_shift = 0 and reg_offset_shift = 1 (For TI)

Also, if we change this line as above, we also need to change below macros:
#define TORRENT_PHY_PCS_COMMON_OFFSET   0xC000 (In Torrent driver)

#define TORRENT_PHY_PMA_COMMON_OFFSET   0xE000 (In Torrent driver)


#define SIERRA_PHY_CONFIG_CTRL_OFFSET	0xc000 (In Sierra driver)

as shifting by block_offset_shift will not be done in cdns_regmap_init().

Are you going to change this implementation in sierra driver? Or if you have already made the changes, can you share them for the reference?

Reply:
yes, I'm making some changes from the previous version.
I've uploaded my wip branch here https://github.com/kishon/linux-wip.git  j7_serdes_v2